src/main.rs

```diff
@@ -69,29 +79,64 @@ impl Client {
     // Method should return an error when a connection already exists.
     // The client should send a handshake to the server.
     fn open(&mut self, addr: &str, server: Server) -> CommsResult<()> {
-        todo!()
+        if self.is_open(addr) {
+            Err(CommsError::ConnectionExists(addr.to_string()))
+        } else {
+            self.connections
+                .insert(addr.to_string(), Connection::Open(server));
+
+            // why not use the send fn if we already have it
+            let handshake = Message {
+                msg_type: MessageType::Handshake,
+                load: self.ip.clone(),
+            };
+
+            match self.send(addr, handshake) {
+                Ok(_) => Ok(()),
+                Err(err) => Err(err), // sadly you cannot do err => err
```

Comment:
Jeśli chcesz zmienić coś tylko w wariancie `Ok` resultu, a błąd zostawić taki sam służą do tego kombinatory:
Jeśli chcemy coś zrobić z wartością w środku `Ok` wtedy używamy `map()` -> `self.send().map(|response| println!("{}", response))`
Jeśli chcemy po prostu zwrócić inne `Ok`, ale `Err` zostawić to samo wtedy używamy `and()` -> `self.send().and(Ok(()))`